esp-hal/src/dma/mod.rs

```diff
@@ -1860,22 +1860,197 @@ pub trait DmaRxBuffer {
 }
 
 /// Error returned from Dma[Rx|Tx|RxTx]Buf operations.
-#[derive(Debug)]
+#[derive(Debug, PartialEq, Clone, Copy)]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
 pub enum DmaBufError {
     /// More descriptors are needed for the buffer size
     InsufficientDescriptors,
     /// Descriptors or buffers are not located in a supported memory region
     UnsupportedMemoryRegion,
 }
 
+#[derive(Debug)]
+struct DescriptorSet<'a> {
+    descriptors: &'a mut [DmaDescriptor],
+}
+
+impl<'a> DescriptorSet<'a> {
+    fn new(descriptors: &'a mut [DmaDescriptor], buffer: &mut [u8]) -> Result<Self, DmaBufError> {
+        if !is_slice_in_dram(descriptors) {
+            return Err(DmaBufError::UnsupportedMemoryRegion);
+        }
+
+        descriptors.fill(DmaDescriptor::EMPTY);
+
+        unsafe { Self::new_from_unchecked(descriptors, buffer) }
+    }
+
+    /// # Safety
+    ///
+    /// The caller must ensure that the descriptors are in a supported memory
+    /// region.
+    unsafe fn new_from_unchecked(
+        descriptors: &'a mut [DmaDescriptor],
+        buffer: &mut [u8],
+    ) -> Result<Self, DmaBufError> {
+        let mut this = Self { descriptors };
+
+        this.link_with_buffer(buffer)?;
+
+        Ok(this)
+    }
+
+    fn scan(&self, mut f: impl FnMut(&DmaDescriptor)) {
+        for desc in self.iter() {
+            f(desc);
+        }
+    }
+
+    fn scan_mut(&mut self, mut f: impl FnMut(&mut DmaDescriptor)) {
+        for desc in self.iter_mut() {
+            f(desc);
+        }
+    }
+
+    fn into_inner(self) -> &'a mut [DmaDescriptor] {
+        self.descriptors
+    }
+
+    fn head(&mut self) -> *mut DmaDescriptor {
+        self.descriptors.as_mut_ptr()
+    }
+
+    /// Prepares descriptors for transferring `len` bytes of data.
+    fn prepare_desciptors(&mut self, len: usize, prepare: fn(&mut DmaDescriptor, usize) -> usize) {
+        // First, pick enough descriptors to cover the buffer.
+        let descriptor_count = len.div_ceil(CHUNK_SIZE).max(1);
```

Comment:
Don't hard code the chunk size in here

Reply:
This was the previous logic, too, I just moved it. Not changing it here, we can deal with it in a follow-up.